{
 "metadata": {
  "name": "",
  "signature": "sha256:2a7ffce0b29e8b84cefc626e6dda585909170551b12ec56e6e7a8b1c2c3d76d4"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "Generate files for motif finding with Homer.\n",
      "\n",
      "Actual motif running done at the command line with Homer.\n",
      "\n",
      "Note that Homer requires the \\r line ending for all input files.\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1,
       "text": [
        "'\\nGenerate files for motif finding with Homer.\\n\\nActual motif running done at the command line with Homer.\\n\\nNote that Homer requires the \\r line ending for all input files.\\n'"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas as pd\n",
      "import matplotlib.pyplot as plt"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ab = 'h3k27ac'\n",
      "data = pd.io.parsers.read_csv('../offstate/data/union_peaks_1um_{}.txt'.format(ab), \n",
      "                                    sep='\\t', \n",
      "                                    header=0)\n",
      "\n",
      "data = data.fillna(0)\n",
      "\n",
      "# Rename our columns of interest\n",
      "labels = ['No Peptide', '1uM K99A', '1uM PCC']\n",
      "last_columns = [labels[1], labels[2], labels[0]]\n",
      "\n",
      "data.columns = data.columns[:-len(last_columns)].tolist() + last_columns\n",
      "\n",
      "data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>PeakID (cmd=annotatePeaks.pl /data/mm10/CD4/ChIP/H3K27Ac/AND-CD4TCell-H3K27Ac-1uM-24h-MNase-All-Conditions/regions.txt mm10 -size given -d AND-CD4TCell-H3K27Ac-1uM-K99A-24h-MNase-Combined AND-CD4TCell-H3K27Ac-1uM-PCC-24h-MNase-Combined AND-CD4TCell-H3K27Ac-No-Peptide-24h-MNase-Combined)</th>\n",
        "      <th>Chr</th>\n",
        "      <th>Start</th>\n",
        "      <th>End</th>\n",
        "      <th>Strand</th>\n",
        "      <th>Peak Score</th>\n",
        "      <th>Focus Ratio/Region Size</th>\n",
        "      <th>Annotation</th>\n",
        "      <th>Detailed Annotation</th>\n",
        "      <th>Distance to TSS</th>\n",
        "      <th>...</th>\n",
        "      <th>Nearest Unigene</th>\n",
        "      <th>Nearest Refseq</th>\n",
        "      <th>Nearest Ensembl</th>\n",
        "      <th>Gene Name</th>\n",
        "      <th>Gene Alias</th>\n",
        "      <th>Gene Description</th>\n",
        "      <th>Gene Type</th>\n",
        "      <th>1uM K99A</th>\n",
        "      <th>1uM PCC</th>\n",
        "      <th>No Peptide</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> chr14-15663</td>\n",
        "      <td> chr14</td>\n",
        "      <td> 54188639</td>\n",
        "      <td> 54237983</td>\n",
        "      <td> +</td>\n",
        "      <td> 4693.6</td>\n",
        "      <td>  36.003</td>\n",
        "      <td>                           Intergenic</td>\n",
        "      <td>                        MIRb|SINE|MIR</td>\n",
        "      <td>  40618</td>\n",
        "      <td>...</td>\n",
        "      <td> Mm.319038</td>\n",
        "      <td> NM_010015</td>\n",
        "      <td> ENSMUSG00000022174</td>\n",
        "      <td>  Dad1</td>\n",
        "      <td>                                AI323713</td>\n",
        "      <td>            defender against cell death 1</td>\n",
        "      <td> protein-coding</td>\n",
        "      <td> 4986.67</td>\n",
        "      <td> 5335.32</td>\n",
        "      <td> 3723.82</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> chr14-18437</td>\n",
        "      <td> chr14</td>\n",
        "      <td> 54143129</td>\n",
        "      <td> 54187413</td>\n",
        "      <td> +</td>\n",
        "      <td> 1923.1</td>\n",
        "      <td>  67.050</td>\n",
        "      <td>                           Intergenic</td>\n",
        "      <td>                           Intergenic</td>\n",
        "      <td>  88658</td>\n",
        "      <td>...</td>\n",
        "      <td> Mm.319038</td>\n",
        "      <td> NM_010015</td>\n",
        "      <td> ENSMUSG00000022174</td>\n",
        "      <td>  Dad1</td>\n",
        "      <td>                                AI323713</td>\n",
        "      <td>            defender against cell death 1</td>\n",
        "      <td> protein-coding</td>\n",
        "      <td> 2070.95</td>\n",
        "      <td> 2199.92</td>\n",
        "      <td> 1452.17</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td>  chr6-17052</td>\n",
        "      <td>  chr6</td>\n",
        "      <td> 41530297</td>\n",
        "      <td> 41560047</td>\n",
        "      <td> +</td>\n",
        "      <td> 1908.1</td>\n",
        "      <td> 127.227</td>\n",
        "      <td>                           Intergenic</td>\n",
        "      <td>                           Intergenic</td>\n",
        "      <td>  23396</td>\n",
        "      <td>...</td>\n",
        "      <td> Mm.276926</td>\n",
        "      <td> NM_009430</td>\n",
        "      <td> ENSMUSG00000057163</td>\n",
        "      <td> Prss2</td>\n",
        "      <td>                 TRY8|TRYP|Ta|Tesp4|Try2</td>\n",
        "      <td>                       protease, serine 2</td>\n",
        "      <td> protein-coding</td>\n",
        "      <td> 2043.98</td>\n",
        "      <td> 2208.75</td>\n",
        "      <td> 1419.29</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td>  chr9-56390</td>\n",
        "      <td>  chr9</td>\n",
        "      <td> 32694702</td>\n",
        "      <td> 32734546</td>\n",
        "      <td> +</td>\n",
        "      <td> 1655.3</td>\n",
        "      <td>  30.703</td>\n",
        "      <td> intron (NM_001038642, intron 1 of 6)</td>\n",
        "      <td> intron (NM_001038642, intron 1 of 6)</td>\n",
        "      <td>  18582</td>\n",
        "      <td>...</td>\n",
        "      <td> Mm.292415</td>\n",
        "      <td> NM_011808</td>\n",
        "      <td> ENSMUSG00000032035</td>\n",
        "      <td>  Ets1</td>\n",
        "      <td> AI196000|AI448617|D230050P06|Ets-1|Tpl1</td>\n",
        "      <td> E26 avian leukemia oncogene 1, 5' domain</td>\n",
        "      <td> protein-coding</td>\n",
        "      <td> 1633.53</td>\n",
        "      <td> 2163.54</td>\n",
        "      <td> 1196.88</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td>  chr9-18001</td>\n",
        "      <td>  chr9</td>\n",
        "      <td> 32904435</td>\n",
        "      <td> 32944129</td>\n",
        "      <td> +</td>\n",
        "      <td> 1576.7</td>\n",
        "      <td>  53.630</td>\n",
        "      <td>                           Intergenic</td>\n",
        "      <td>                           Intergenic</td>\n",
        "      <td> 228240</td>\n",
        "      <td>...</td>\n",
        "      <td> Mm.292415</td>\n",
        "      <td> NM_011808</td>\n",
        "      <td> ENSMUSG00000032035</td>\n",
        "      <td>  Ets1</td>\n",
        "      <td> AI196000|AI448617|D230050P06|Ets-1|Tpl1</td>\n",
        "      <td> E26 avian leukemia oncogene 1, 5' domain</td>\n",
        "      <td> protein-coding</td>\n",
        "      <td> 1720.98</td>\n",
        "      <td> 1725.41</td>\n",
        "      <td> 1256.14</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 22 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 15,
       "text": [
        "  PeakID (cmd=annotatePeaks.pl /data/mm10/CD4/ChIP/H3K27Ac/AND-CD4TCell-H3K27Ac-1uM-24h-MNase-All-Conditions/regions.txt mm10 -size given -d AND-CD4TCell-H3K27Ac-1uM-K99A-24h-MNase-Combined AND-CD4TCell-H3K27Ac-1uM-PCC-24h-MNase-Combined AND-CD4TCell-H3K27Ac-No-Peptide-24h-MNase-Combined)  \\\n",
        "0                                        chr14-15663                                                                                                                                                                                                                                                \n",
        "1                                        chr14-18437                                                                                                                                                                                                                                                \n",
        "2                                         chr6-17052                                                                                                                                                                                                                                                \n",
        "3                                         chr9-56390                                                                                                                                                                                                                                                \n",
        "4                                         chr9-18001                                                                                                                                                                                                                                                \n",
        "\n",
        "     Chr     Start       End Strand  Peak Score  Focus Ratio/Region Size  \\\n",
        "0  chr14  54188639  54237983      +      4693.6                   36.003   \n",
        "1  chr14  54143129  54187413      +      1923.1                   67.050   \n",
        "2   chr6  41530297  41560047      +      1908.1                  127.227   \n",
        "3   chr9  32694702  32734546      +      1655.3                   30.703   \n",
        "4   chr9  32904435  32944129      +      1576.7                   53.630   \n",
        "\n",
        "                             Annotation                   Detailed Annotation  \\\n",
        "0                            Intergenic                         MIRb|SINE|MIR   \n",
        "1                            Intergenic                            Intergenic   \n",
        "2                            Intergenic                            Intergenic   \n",
        "3  intron (NM_001038642, intron 1 of 6)  intron (NM_001038642, intron 1 of 6)   \n",
        "4                            Intergenic                            Intergenic   \n",
        "\n",
        "   Distance to TSS     ...     Nearest Unigene  Nearest Refseq  \\\n",
        "0            40618     ...           Mm.319038       NM_010015   \n",
        "1            88658     ...           Mm.319038       NM_010015   \n",
        "2            23396     ...           Mm.276926       NM_009430   \n",
        "3            18582     ...           Mm.292415       NM_011808   \n",
        "4           228240     ...           Mm.292415       NM_011808   \n",
        "\n",
        "      Nearest Ensembl Gene Name                               Gene Alias  \\\n",
        "0  ENSMUSG00000022174      Dad1                                 AI323713   \n",
        "1  ENSMUSG00000022174      Dad1                                 AI323713   \n",
        "2  ENSMUSG00000057163     Prss2                  TRY8|TRYP|Ta|Tesp4|Try2   \n",
        "3  ENSMUSG00000032035      Ets1  AI196000|AI448617|D230050P06|Ets-1|Tpl1   \n",
        "4  ENSMUSG00000032035      Ets1  AI196000|AI448617|D230050P06|Ets-1|Tpl1   \n",
        "\n",
        "                           Gene Description       Gene Type 1uM K99A  1uM PCC  \\\n",
        "0             defender against cell death 1  protein-coding  4986.67  5335.32   \n",
        "1             defender against cell death 1  protein-coding  2070.95  2199.92   \n",
        "2                        protease, serine 2  protein-coding  2043.98  2208.75   \n",
        "3  E26 avian leukemia oncogene 1, 5' domain  protein-coding  1633.53  2163.54   \n",
        "4  E26 avian leukemia oncogene 1, 5' domain  protein-coding  1720.98  1725.41   \n",
        "\n",
        "   No Peptide  \n",
        "0     3723.82  \n",
        "1     1452.17  \n",
        "2     1419.29  \n",
        "3     1196.88  \n",
        "4     1256.14  \n",
        "\n",
        "[5 rows x 22 columns]"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Enhancers only\n",
      "data = data[data['Annotation'].str.contains('promoter') == False]\n",
      "#data = data[data['Distance to TSS'] < 5000]\n",
      "data.columns"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 16,
       "text": [
        "Index(['PeakID (cmd=annotatePeaks.pl /data/mm10/CD4/ChIP/H3K27Ac/AND-CD4TCell-H3K27Ac-1uM-24h-MNase-All-Conditions/regions.txt mm10 -size given -d AND-CD4TCell-H3K27Ac-1uM-K99A-24h-MNase-Combined AND-CD4TCell-H3K27Ac-1uM-PCC-24h-MNase-Combined AND-CD4TCell-H3K27Ac-No-Peptide-24h-MNase-Combined)', 'Chr', 'Start', 'End', 'Strand', 'Peak Score', 'Focus Ratio/Region Size', 'Annotation', 'Detailed Annotation', 'Distance to TSS', 'Nearest PromoterID', 'Entrez ID', 'Nearest Unigene', 'Nearest Refseq', 'Nearest Ensembl', 'Gene Name', 'Gene Alias', 'Gene Description', 'Gene Type', '1uM K99A', '1uM PCC', 'No Peptide'], dtype='object')"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "min_tags = 10"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Activated only?\n",
      "fold = 2\n",
      "act = labels[1:]\n",
      "act_only = data[(data[act].max(axis=1) > data[labels[0]]*fold) &\n",
      "              (data[act].max(axis=1) > min_tags)]\n",
      "print('Activated but not {} enhancers'.format(labels[0]), len(act_only))\n",
      "act_only[labels].describe()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Activated but not No Peptide enhancers 1257\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>No Peptide</th>\n",
        "      <th>1uM K99A</th>\n",
        "      <th>1uM PCC</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>count</th>\n",
        "      <td> 1257.000000</td>\n",
        "      <td> 1257.000000</td>\n",
        "      <td> 1257.000000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mean</th>\n",
        "      <td>   25.810605</td>\n",
        "      <td>   34.555163</td>\n",
        "      <td>   62.427192</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>std</th>\n",
        "      <td>   35.646348</td>\n",
        "      <td>   47.831617</td>\n",
        "      <td>   82.444186</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>min</th>\n",
        "      <td>    1.880000</td>\n",
        "      <td>    2.060000</td>\n",
        "      <td>    4.520000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>25%</th>\n",
        "      <td>    7.120000</td>\n",
        "      <td>    9.960000</td>\n",
        "      <td>   18.300000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>50%</th>\n",
        "      <td>   12.150000</td>\n",
        "      <td>   16.670000</td>\n",
        "      <td>   31.000000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>75%</th>\n",
        "      <td>   27.230000</td>\n",
        "      <td>   36.080000</td>\n",
        "      <td>   67.820000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>max</th>\n",
        "      <td>  332.990000</td>\n",
        "      <td>  445.840000</td>\n",
        "      <td>  699.940000</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 20,
       "text": [
        "        No Peptide     1uM K99A      1uM PCC\n",
        "count  1257.000000  1257.000000  1257.000000\n",
        "mean     25.810605    34.555163    62.427192\n",
        "std      35.646348    47.831617    82.444186\n",
        "min       1.880000     2.060000     4.520000\n",
        "25%       7.120000     9.960000    18.300000\n",
        "50%      12.150000    16.670000    31.000000\n",
        "75%      27.230000    36.080000    67.820000\n",
        "max     332.990000   445.840000   699.940000"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "act_only[['Chr', 'Start', 'End', 'Strand']].to_csv(\n",
      "            'output/activated_enhancers_{}x_{}.txt'.format(fold, ab), \n",
      "              sep='\\t', header=True, index=True,\n",
      "              line_terminator='\\r\\n')\n",
      "data[['Chr', 'Start', 'End', 'Strand']].to_csv(\n",
      "            'output/all_enhancers_{}.txt'.format(ab), \n",
      "              sep='\\t', header=True, index=True,\n",
      "              line_terminator='\\r\\n')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "Then run Homer:\n",
      "\n",
      "for f in *.txt; \n",
      "do \n",
      "OUTPUT=`echo ${f} | sed 's/\\.txt//g'`\n",
      "findMotifsGenome.pl $f mm10 ${OUTPUT}_motifs -size -100,+100 -len 8,10,12 -p 6;\n",
      "done;\n",
      "\n",
      "for f in *.txt; \n",
      "do \n",
      "OUTPUT=`echo ${f} | sed 's/\\.txt//g'`\n",
      "findMotifsGenome.pl $f mm10 ${OUTPUT}_known -mknown ../../motifs_of_interest/motifs_of_interest.motifs -nomotif -size -100,+100 -p 6\n",
      "done;\n",
      "\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "\"\\nThen run Homer:\\n\\nfor f in *.txt; \\ndo \\nOUTPUT=`echo ${f} | sed 's/\\\\.txt//g'`\\nfindMotifsGenome.pl $f mm10 ${OUTPUT}_motifs -size -100,+100 -len 8,10,12 -p 6;\\ndone;\\n\\nfor f in *.txt; \\ndo \\nOUTPUT=`echo ${f} | sed 's/\\\\.txt//g'`\\nfindMotifsGenome.pl $f mm10 ${OUTPUT}_known -mknown ../../motifs_of_interest/motifs_of_interest.motifs -nomotif -size -100,+100 -p 6\\ndone;\\n\\n\""
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "For parsing known results:\n",
      "\n",
      "mkdir 1um_known_results\n",
      "for f in *known/knownResults.txt; \n",
      "do \n",
      "OUTPUT=`dirname ${f}`\n",
      "cp $f activation_specific_known_results/${OUTPUT}.txt;\n",
      "done;\n",
      "\n",
      "'''\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 9,
       "text": [
        "'\\nFor parsing known results:\\n\\nmkdir 1um_all_known_results 1um_only_known_results\\nfor f in *all*known/knownResults.txt; \\ndo \\nOUTPUT=`dirname ${f}`\\ncp $f 1um_all_known_results/${OUTPUT}.txt;\\ndone;\\n\\nfor f in *only*known/knownResults.txt; \\ndo \\nOUTPUT=`dirname ${f}`\\ncp $f 1um_only_known_results/${OUTPUT}.txt;\\ndone;\\n'"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "motifs = ['1-ACAGGAAGTG,BestGuess:ERG(ETS)/VCaP-ERG-ChIP-Seq(GSE14097)/Homer(0.984)',\n",
      "          'RUNX1(Runt)/Jurkat-RUNX1-ChIP-Seq(GSE29180)/Homer',\n",
      "          'IRF2(IRF)/Erythroblas-IRF2-ChIP-Seq(GSE36985)/Homer',\n",
      "          'AP-1(bZIP)/ThioMac-PU.1-ChIP-Seq(GSE21512)/Homer',\n",
      "          'Klf4(Zf)/mES-Klf4-ChIP-Seq(GSE11431)/Homer',\n",
      "          'NFkB-p65-Rel(RHD)/LPS-exp(GSE23622)/Homer']\n",
      "names = ['ETS', 'Runx', 'IRF', 'AP-1', 'KLF', 'NF-kB']\n",
      "labels = ['Activation-specific Enhancers', 'All Enhancers']\n",
      "files = ['activated_enhancers_{}x_H3K4me2'.format(fold), 'all_enhancers_H3K4me2']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}