{
 "metadata": {
  "name": "",
  "signature": "sha256:e1a4e43bd067c63726ec7f2c3785f56947217fe59d7938388cdaef27ab9e8ac4"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "Use RefSeq genes to get AP-1 counts (public Th17 data) per gene.\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 6,
       "text": [
        "'\\nUse RefSeq genes to get AP-1 counts (public Th17 data) per gene.\\n'"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas.io.sql import read_sql_query\n",
      "from sqlalchemy import create_engine\n",
      "\n",
      "\n",
      "def get_engine(uri, password, username='vespucci_user', dbname='vespucci'):\n",
      "    '''\n",
      "    dialect+driver://username:password@host:port/database\n",
      "    '''\n",
      "    return create_engine('postgresql://{}:{}@{}:5432/{}'.format(\n",
      "        username, password, uri, dbname))\n",
      "\n",
      "\n",
      "def dataframe_from_query(query, engine):\n",
      "    '''\n",
      "    Given the database connection and query, return a Pandas dataframe\n",
      "    with the requested data.\n",
      "    '''\n",
      "    return read_sql_query(query, engine)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "engine = get_engine(\n",
      "    uri='ec2-54-160-104-39.compute-1.amazonaws.com',\n",
      "    password='vespucci 2015')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sql = '''\n",
      "SELECT\n",
      "seq.sequence_identifier,\n",
      "chrom.\"name\" as chr_name,\n",
      "reg.transcription_start,\n",
      "reg.transcription_end,\n",
      "reg.strand,\n",
      "peak.tag_count as {peak}_tag_count\n",
      "\n",
      "FROM genome_reference_mm10.sequence_identifier seq\n",
      "JOIN genome_reference_mm10.sequence_transcription_region reg\n",
      "ON seq.id = reg.sequence_identifier_id\n",
      "JOIN genome_reference_mm10.chromosome chrom\n",
      "ON reg.chromosome_id = chrom.id\n",
      "LEFT OUTER JOIN chipseq.peak_cd4tcell_th17_{peak} peak\n",
      "ON reg.chromosome_id = peak.chromosome_id\n",
      "AND reg.start_end_1000 && peak.start_end\n",
      "\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "datasets = []\n",
      "peaks = ['batf', 'cjun', 'irf4', 'junb', 'jund']\n",
      "for peak in peaks:\n",
      "    data = dataframe_from_query(sql.format(peak=peak), engine)\n",
      "    datasets.append(data)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "all_data = pd.concat(datasets, join='outer', axis=0, ignore_index=True)\n",
      "all_data = all_data.fillna(0)\n",
      "all_data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>batf_tag_count</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>cjun_tag_count</th>\n",
        "      <th>irf4_tag_count</th>\n",
        "      <th>junb_tag_count</th>\n",
        "      <th>jund_tag_count</th>\n",
        "      <th>sequence_identifier</th>\n",
        "      <th>strand</th>\n",
        "      <th>transcription_end</th>\n",
        "      <th>transcription_start</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> 17.2</td>\n",
        "      <td>  chrX</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    NM_181516</td>\n",
        "      <td> 0</td>\n",
        "      <td>  74290151</td>\n",
        "      <td>  74282717</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 15.3</td>\n",
        "      <td>  chr2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    NM_019806</td>\n",
        "      <td> 0</td>\n",
        "      <td> 173784336</td>\n",
        "      <td> 173737570</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td>  0.0</td>\n",
        "      <td>  chr6</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    NM_027868</td>\n",
        "      <td> 0</td>\n",
        "      <td>  90646412</td>\n",
        "      <td>  90604879</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 11.5</td>\n",
        "      <td> chr13</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    NM_029001</td>\n",
        "      <td> 0</td>\n",
        "      <td> 108287107</td>\n",
        "      <td> 108214403</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 39.6</td>\n",
        "      <td>  chr9</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> NM_001042707</td>\n",
        "      <td> 0</td>\n",
        "      <td>  21405363</td>\n",
        "      <td>  21367870</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 6,
       "text": [
        "   batf_tag_count chr_name  cjun_tag_count  irf4_tag_count  junb_tag_count  \\\n",
        "0            17.2     chrX               0               0               0   \n",
        "1            15.3     chr2               0               0               0   \n",
        "2             0.0     chr6               0               0               0   \n",
        "3            11.5    chr13               0               0               0   \n",
        "4            39.6     chr9               0               0               0   \n",
        "\n",
        "   jund_tag_count sequence_identifier  strand  transcription_end  \\\n",
        "0               0           NM_181516       0           74290151   \n",
        "1               0           NM_019806       0          173784336   \n",
        "2               0           NM_027868       0           90646412   \n",
        "3               0           NM_029001       0          108287107   \n",
        "4               0        NM_001042707       0           21405363   \n",
        "\n",
        "   transcription_start  \n",
        "0             74282717  \n",
        "1            173737570  \n",
        "2             90604879  \n",
        "3            108214403  \n",
        "4             21367870  "
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# The query will return one row for each overlapping peak\n",
      "# Collapse into one row per refseq id\n",
      "grouped = all_data.groupby(['sequence_identifier',\n",
      "                 'chr_name',\n",
      "                 'transcription_start',\n",
      "                 'transcription_end',\n",
      "                 'strand']).sum()\n",
      "grouped.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th>batf_tag_count</th>\n",
        "      <th>cjun_tag_count</th>\n",
        "      <th>irf4_tag_count</th>\n",
        "      <th>junb_tag_count</th>\n",
        "      <th>jund_tag_count</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>sequence_identifier</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>transcription_start</th>\n",
        "      <th>transcription_end</th>\n",
        "      <th>strand</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>NM_001001130</th>\n",
        "      <th>chr13</th>\n",
        "      <th>67747799 </th>\n",
        "      <th>67755134 </th>\n",
        "      <th>1</th>\n",
        "      <td> 15.30</td>\n",
        "      <td>  8.87</td>\n",
        "      <td> 33.60</td>\n",
        "      <td>  0.0</td>\n",
        "      <td> 16.1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>NM_001001144</th>\n",
        "      <th>chr9</th>\n",
        "      <th>110333355</th>\n",
        "      <th>110384949</th>\n",
        "      <th>0</th>\n",
        "      <td> 26.84</td>\n",
        "      <td> 14.30</td>\n",
        "      <td> 35.70</td>\n",
        "      <td> 27.6</td>\n",
        "      <td> 63.5</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>NM_001001152</th>\n",
        "      <th>chr13</th>\n",
        "      <th>67254917 </th>\n",
        "      <th>67269068 </th>\n",
        "      <th>1</th>\n",
        "      <td> 27.50</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  6.69</td>\n",
        "      <td>  0.0</td>\n",
        "      <td> 40.6</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>NM_001001160</th>\n",
        "      <th>chr6</th>\n",
        "      <th>85469575 </th>\n",
        "      <th>85502994 </th>\n",
        "      <th>1</th>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.0</td>\n",
        "      <td>  7.8</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>NM_001001176</th>\n",
        "      <th>chrX</th>\n",
        "      <th>106206873</th>\n",
        "      <th>106219842</th>\n",
        "      <th>1</th>\n",
        "      <td>  0.00</td>\n",
        "      <td> 20.50</td>\n",
        "      <td> 10.40</td>\n",
        "      <td>  0.0</td>\n",
        "      <td> 29.1</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "                                                                           batf_tag_count  \\\n",
        "sequence_identifier chr_name transcription_start transcription_end strand                   \n",
        "NM_001001130        chr13    67747799            67755134          1                15.30   \n",
        "NM_001001144        chr9     110333355           110384949         0                26.84   \n",
        "NM_001001152        chr13    67254917            67269068          1                27.50   \n",
        "NM_001001160        chr6     85469575            85502994          1                 0.00   \n",
        "NM_001001176        chrX     106206873           106219842         1                 0.00   \n",
        "\n",
        "                                                                           cjun_tag_count  \\\n",
        "sequence_identifier chr_name transcription_start transcription_end strand                   \n",
        "NM_001001130        chr13    67747799            67755134          1                 8.87   \n",
        "NM_001001144        chr9     110333355           110384949         0                14.30   \n",
        "NM_001001152        chr13    67254917            67269068          1                 0.00   \n",
        "NM_001001160        chr6     85469575            85502994          1                 0.00   \n",
        "NM_001001176        chrX     106206873           106219842         1                20.50   \n",
        "\n",
        "                                                                           irf4_tag_count  \\\n",
        "sequence_identifier chr_name transcription_start transcription_end strand                   \n",
        "NM_001001130        chr13    67747799            67755134          1                33.60   \n",
        "NM_001001144        chr9     110333355           110384949         0                35.70   \n",
        "NM_001001152        chr13    67254917            67269068          1                 6.69   \n",
        "NM_001001160        chr6     85469575            85502994          1                 0.00   \n",
        "NM_001001176        chrX     106206873           106219842         1                10.40   \n",
        "\n",
        "                                                                           junb_tag_count  \\\n",
        "sequence_identifier chr_name transcription_start transcription_end strand                   \n",
        "NM_001001130        chr13    67747799            67755134          1                  0.0   \n",
        "NM_001001144        chr9     110333355           110384949         0                 27.6   \n",
        "NM_001001152        chr13    67254917            67269068          1                  0.0   \n",
        "NM_001001160        chr6     85469575            85502994          1                  0.0   \n",
        "NM_001001176        chrX     106206873           106219842         1                  0.0   \n",
        "\n",
        "                                                                           jund_tag_count  \n",
        "sequence_identifier chr_name transcription_start transcription_end strand                  \n",
        "NM_001001130        chr13    67747799            67755134          1                 16.1  \n",
        "NM_001001144        chr9     110333355           110384949         0                 63.5  \n",
        "NM_001001152        chr13    67254917            67269068          1                 40.6  \n",
        "NM_001001160        chr6     85469575            85502994          1                  7.8  \n",
        "NM_001001176        chrX     106206873           106219842         1                 29.1  "
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "grouped.to_csv('../data/ap1_per_gene.txt', \n",
      "            sep='\\t', header=True, index=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}