{
 "metadata": {
  "name": "",
  "signature": "sha256:d094b89c60dbbe07f68b32651a421a619ec4ced0eb960bca4be78e0ddbf99199"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import OrderedDict\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas.io.sql import read_sql_query\n",
      "from sqlalchemy import create_engine\n",
      "\n",
      "\n",
      "def get_engine(uri, password, username='vespucci_user', dbname='vespucci'):\n",
      "    '''\n",
      "    dialect+driver://username:password@host:port/database\n",
      "    '''\n",
      "    return create_engine('postgresql://{}:{}@{}:5432/{}'.format(\n",
      "        username, password, uri, dbname))\n",
      "\n",
      "\n",
      "def dataframe_from_query(query, engine):\n",
      "    '''\n",
      "    Given the database connection and query, return a Pandas dataframe\n",
      "    with the requested data.\n",
      "    '''\n",
      "    return read_sql_query(query, engine)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "engine = get_engine(\n",
      "    uri='ec2-54-160-104-39.compute-1.amazonaws.com',\n",
      "    password='vespucci 2015')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ab = 'h3k4me2'\n",
      "peptides = ['no_peptide', '1um_k99a', '1um_pcc']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sql = '''select distinct on (p1.id) \n",
      "p1.id, chr.\"name\" as chr_name, p1.start, p1.\"end\",\n",
      "p1.chromosome_id, \n",
      "p1.tag_count as tag_count_{peptide},\n",
      "p1.id as id_{peptide},\n",
      "p2.tag_count as tag_count_{other_1},\n",
      "p2.id as id_{other_1},\n",
      "p3.tag_count as tag_count_{other_2},\n",
      "p3.id as id_{other_2}\n",
      "\n",
      "from chipseq.peak_cd4tcell_{ab}_{peptide}_24h p1\n",
      "join genome_reference_mm10.chromosome chr \n",
      "on p1.chromosome_id = chr.id\n",
      "\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_1}_24h p2\n",
      "on p1.chromosome_id = p2.chromosome_id\n",
      "and p1.start_end && p2.start_end\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_2}_24h p3\n",
      "on p1.chromosome_id = p3.chromosome_id\n",
      "and p1.start_end && p3.start_end\n",
      "\n",
      "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
      "on p1.chromosome_id = reg.chromosome_id\n",
      "and p1.start_end && reg.start_site_1000\n",
      "\n",
      "where reg.id is {genic} NULL;\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "others_labels = ['other_1', 'other_2']\n",
      "for i, current in enumerate(peptides):\n",
      "    others = peptides[:i] + peptides[i+1:]\n",
      "    others_d = OrderedDict(zip(others_labels, others))\n",
      "    \n",
      "    nongene_sql = sql.format(ab=ab, peptide=current, genic='', **others_d)\n",
      "    gene_sql = sql.format(ab=ab, peptide=current, genic='NOT', **others_d)\n",
      "    \n",
      "    data = dataframe_from_query(nongene_sql, engine)\n",
      "    data.to_csv('data/enhancers_me2_and_{}.txt'.format(current), \n",
      "                sep='\\t', header=True, index=False)\n",
      "    data = dataframe_from_query(gene_sql, engine)\n",
      "    data.to_csv('data/promoters_me2_and_{}.txt'.format(current),\n",
      "                sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>id</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>start</th>\n",
        "      <th>end</th>\n",
        "      <th>chromosome_id</th>\n",
        "      <th>tag_count_10um_pcc</th>\n",
        "      <th>id_10um_pcc</th>\n",
        "      <th>tag_count_no_peptide</th>\n",
        "      <th>id_no_peptide</th>\n",
        "      <th>tag_count_10um_k99a</th>\n",
        "      <th>...</th>\n",
        "      <th>batf_tag_count</th>\n",
        "      <th>batf_id</th>\n",
        "      <th>junb_tag_count</th>\n",
        "      <th>junb_id</th>\n",
        "      <th>cjun_tag_count</th>\n",
        "      <th>cjun_id</th>\n",
        "      <th>jund_tag_count</th>\n",
        "      <th>jund_id</th>\n",
        "      <th>irf4_tag_count</th>\n",
        "      <th>irf4_id</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>  5</td>\n",
        "      <td> chr11</td>\n",
        "      <td>  22838517</td>\n",
        "      <td>  22859099</td>\n",
        "      <td> 13</td>\n",
        "      <td> 1045.5</td>\n",
        "      <td>  5</td>\n",
        "      <td>  475.6</td>\n",
        "      <td>  207</td>\n",
        "      <td> 1217.1</td>\n",
        "      <td>...</td>\n",
        "      <td> 53.70</td>\n",
        "      <td>  5125</td>\n",
        "      <td> 19.2</td>\n",
        "      <td> 17391</td>\n",
        "      <td>  NaN</td>\n",
        "      <td>   NaN</td>\n",
        "      <td>  35.0</td>\n",
        "      <td> 20005</td>\n",
        "      <td> 24.2</td>\n",
        "      <td> 19367</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td>  8</td>\n",
        "      <td> chr15</td>\n",
        "      <td> 101252619</td>\n",
        "      <td> 101272613</td>\n",
        "      <td> 16</td>\n",
        "      <td>  887.8</td>\n",
        "      <td>  8</td>\n",
        "      <td>  226.9</td>\n",
        "      <td> 1163</td>\n",
        "      <td>  824.8</td>\n",
        "      <td>...</td>\n",
        "      <td>  9.58</td>\n",
        "      <td> 31653</td>\n",
        "      <td>  NaN</td>\n",
        "      <td>   NaN</td>\n",
        "      <td> 10.2</td>\n",
        "      <td> 24742</td>\n",
        "      <td>  72.1</td>\n",
        "      <td>  5394</td>\n",
        "      <td> 28.6</td>\n",
        "      <td> 15560</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td>  9</td>\n",
        "      <td> chr13</td>\n",
        "      <td>  30746035</td>\n",
        "      <td>  30764187</td>\n",
        "      <td> 14</td>\n",
        "      <td>  839.9</td>\n",
        "      <td>  9</td>\n",
        "      <td>  326.2</td>\n",
        "      <td>  529</td>\n",
        "      <td>  354.5</td>\n",
        "      <td>...</td>\n",
        "      <td> 53.70</td>\n",
        "      <td>  5136</td>\n",
        "      <td> 32.6</td>\n",
        "      <td> 13136</td>\n",
        "      <td> 47.1</td>\n",
        "      <td>  6284</td>\n",
        "      <td>  40.6</td>\n",
        "      <td> 16077</td>\n",
        "      <td> 45.2</td>\n",
        "      <td>  7427</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 10</td>\n",
        "      <td>  chr9</td>\n",
        "      <td> 109049145</td>\n",
        "      <td> 109060763</td>\n",
        "      <td> 12</td>\n",
        "      <td>  814.5</td>\n",
        "      <td> 10</td>\n",
        "      <td> 1081.3</td>\n",
        "      <td>   11</td>\n",
        "      <td>  985.8</td>\n",
        "      <td>...</td>\n",
        "      <td>   NaN</td>\n",
        "      <td>   NaN</td>\n",
        "      <td> 47.6</td>\n",
        "      <td>  9476</td>\n",
        "      <td> 86.7</td>\n",
        "      <td>  1541</td>\n",
        "      <td>  81.8</td>\n",
        "      <td>  3964</td>\n",
        "      <td> 88.3</td>\n",
        "      <td>  1701</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 11</td>\n",
        "      <td>  chr1</td>\n",
        "      <td>  91051660</td>\n",
        "      <td>  91065868</td>\n",
        "      <td>  1</td>\n",
        "      <td>  813.3</td>\n",
        "      <td> 11</td>\n",
        "      <td> 1009.5</td>\n",
        "      <td>   16</td>\n",
        "      <td>  936.1</td>\n",
        "      <td>...</td>\n",
        "      <td> 69.00</td>\n",
        "      <td>  3444</td>\n",
        "      <td> 65.2</td>\n",
        "      <td>  6470</td>\n",
        "      <td> 39.6</td>\n",
        "      <td>  8351</td>\n",
        "      <td> 108.7</td>\n",
        "      <td>  1777</td>\n",
        "      <td> 45.9</td>\n",
        "      <td>  7189</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 27 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 9,
       "text": [
        "   id chr_name      start        end  chromosome_id  tag_count_10um_pcc  \\\n",
        "0   5    chr11   22838517   22859099             13              1045.5   \n",
        "1   8    chr15  101252619  101272613             16               887.8   \n",
        "2   9    chr13   30746035   30764187             14               839.9   \n",
        "3  10     chr9  109049145  109060763             12               814.5   \n",
        "4  11     chr1   91051660   91065868              1               813.3   \n",
        "\n",
        "   id_10um_pcc  tag_count_no_peptide  id_no_peptide  tag_count_10um_k99a  \\\n",
        "0            5                 475.6            207               1217.1   \n",
        "1            8                 226.9           1163                824.8   \n",
        "2            9                 326.2            529                354.5   \n",
        "3           10                1081.3             11                985.8   \n",
        "4           11                1009.5             16                936.1   \n",
        "\n",
        "    ...     batf_tag_count  batf_id  junb_tag_count  junb_id  cjun_tag_count  \\\n",
        "0   ...              53.70     5125            19.2    17391             NaN   \n",
        "1   ...               9.58    31653             NaN      NaN            10.2   \n",
        "2   ...              53.70     5136            32.6    13136            47.1   \n",
        "3   ...                NaN      NaN            47.6     9476            86.7   \n",
        "4   ...              69.00     3444            65.2     6470            39.6   \n",
        "\n",
        "   cjun_id  jund_tag_count  jund_id  irf4_tag_count  irf4_id  \n",
        "0      NaN            35.0    20005            24.2    19367  \n",
        "1    24742            72.1     5394            28.6    15560  \n",
        "2     6284            40.6    16077            45.2     7427  \n",
        "3     1541            81.8     3964            88.3     1701  \n",
        "4     8351           108.7     1777            45.9     7189  \n",
        "\n",
        "[5 rows x 27 columns]"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# We want one unified set of all the peaks,\n",
      "# but there is overlap between each of our sets.\n",
      "# Take the record from the set with the biggest peak\n",
      "# in each case where there are duplicates.\n",
      "# This is not perfect, especially for the cases\n",
      "# where one large peak covers several smaller,\n",
      "# but it gives us a place to start.\n",
      "\n",
      "for peak_type in ('enhancers','promoters'):\n",
      "    keep_dfs = []\n",
      "    for i, pep in enumerate(peptides):\n",
      "        data = pd.io.parsers.read_csv('data/{}_me2_and_{}.txt'.format(peak_type, pep), \n",
      "                                    sep='\\t', header=0)\n",
      "        data = data.fillna(0)\n",
      "        print(len(data))\n",
      "        key = 'tag_count_' + pep\n",
      "\n",
      "        peptides_before = ['tag_count_' + o for o in peptides[:i]]\n",
      "        peptides_after = ['tag_count_' + o for o in peptides[i+1:]]\n",
      "        others = peptides_before + peptides_after\n",
      "        data_mask = data[key] > data[others].max(axis=1)\n",
      "        print(sum(data_mask))\n",
      "\n",
      "        # If there is a tie, only take this row if it is the first instance of the tie.\n",
      "        tie_mask = data[key] == data[others].max(axis=1)\n",
      "        # Is this the first instance? Fill nulls for the case of the first column.\n",
      "        first_mask = data[key] > data[peptides_before].max(axis=1).fillna(0)\n",
      "        print(sum(tie_mask & first_mask))\n",
      "        keep_dfs.append(data[data_mask | (tie_mask & first_mask)])\n",
      "\n",
      "    keep = pd.concat(keep_dfs, join='outer', axis=0, ignore_index=True)\n",
      "    print(len(keep))\n",
      "    keep.head()\n",
      "    keep.to_csv('data/{}_me2_and_union.txt'.format(peak_type), \n",
      "                sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "19472\n",
        "8932\n",
        "54\n",
        "20586"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "6708\n",
        "30\n",
        "19871"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "4489\n",
        "14\n",
        "21207"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "8805\n",
        "5\n",
        "21078"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "6552\n",
        "0\n",
        "35589\n",
        "13381"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "4400\n",
        "29\n",
        "13690"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "4003\n",
        "19\n",
        "12977"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "267\n",
        "3\n",
        "14009"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "4315\n",
        "4\n",
        "13637"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "659\n",
        "0\n",
        "13699\n"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}