{
 "metadata": {
  "name": "",
  "signature": "sha256:4d579fc49ac01dbea6de37d92d8d54f438af76415929505b5772334da5c7c06c"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas.io.sql import read_sql_query\n",
      "from sqlalchemy import create_engine\n",
      "\n",
      "\n",
      "def get_engine(uri, password, username='vespucci_user', dbname='vespucci'):\n",
      "    '''\n",
      "    dialect+driver://username:password@host:port/database\n",
      "    '''\n",
      "    return create_engine('postgresql://{}:{}@{}:5432/{}'.format(\n",
      "        username, password, uri, dbname))\n",
      "\n",
      "\n",
      "def dataframe_from_query(query, engine):\n",
      "    '''\n",
      "    Given the database connection and query, return a Pandas dataframe\n",
      "    with the requested data.\n",
      "    '''\n",
      "    return read_sql_query(query, engine)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "engine = get_engine(\n",
      "    uri='ec2-54-160-104-39.compute-1.amazonaws.com',\n",
      "    password='vespucci 2015')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_sql = '''\n",
      "SELECT\n",
      "DISTINCT ON(t.id)\n",
      "t.id,\n",
      "chr.name as chr_name,\n",
      "t.transcription_start,\n",
      "t.transcription_end,\n",
      "t.strand,\n",
      "t.score,\n",
      "t.rpkm,\n",
      "t.chromosome_id,\n",
      "seq.sequence_identifier,\n",
      "notx.tag_count as notx_tag_count,\n",
      "kla_1h.tag_count as kla_1h_tag_count,\n",
      "kla_6h.tag_count as kla_6h_tag_count,\n",
      "kla_12h.tag_count as kla_12h_tag_count,\n",
      "kla_24h.tag_count as kla_24h_tag_count,\n",
      "notx_run.total_tags as notx_total_tags,\n",
      "kla_1h_run.total_tags as kla_1h_total_tags,\n",
      "kla_6h_run.total_tags as kla_6h_total_tags,\n",
      "kla_12h_run.total_tags as kla_12h_total_tags,\n",
      "kla_24h_run.total_tags as kla_24h_total_tags\n",
      "\n",
      "FROM atlas_mm10_default.atlas_transcript t\n",
      "join genome_reference_mm10.chromosome chr\n",
      "on t.chromosome_id = chr.id\n",
      "\n",
      "-- Gene region and name\n",
      "join atlas_mm10_default.atlas_transcript_sequence tseq\n",
      "on t.id = tseq.atlas_transcript_id\n",
      "join genome_reference_mm10.sequence_transcription_region reg\n",
      "on tseq.sequence_transcription_region_id = reg.id\n",
      "join genome_reference_mm10.sequence_identifier seq\n",
      "on reg.sequence_identifier_id = seq.id\n",
      "\n",
      "-- Tag counts\n",
      "join atlas_mm10_default.atlas_transcript_source notx\n",
      "on t.id = notx.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run notx_run\n",
      "on notx.sequencing_run_id = notx_run.id\n",
      "and notx_run.source_table = 'groseq\".\"tag_thiomac_gro_notx_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_1h\n",
      "on t.id = kla_1h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_1h_run\n",
      "on kla_1h.sequencing_run_id = kla_1h_run.id\n",
      "and kla_1h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_1h_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_6h\n",
      "on t.id = kla_6h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_6h_run\n",
      "on kla_6h.sequencing_run_id = kla_6h_run.id\n",
      "and kla_6h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_6h_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_12h\n",
      "on t.id = kla_12h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_12h_run\n",
      "on kla_12h.sequencing_run_id = kla_12h_run.id\n",
      "and kla_12h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_12h_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_24h\n",
      "on t.id = kla_24h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_24h_run\n",
      "on kla_24h.sequencing_run_id = kla_24h_run.id\n",
      "and kla_24h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_24h_1'\n",
      "\n",
      "where\n",
      "t.refseq = true\n",
      "and tseq.major = true\n",
      "and t.score >= 1\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data = dataframe_from_query(gene_sql, engine)\n",
      "\n",
      "data.to_csv('data/refseq_genes_kla.txt', sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>id</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>transcription_start</th>\n",
        "      <th>transcription_end</th>\n",
        "      <th>strand</th>\n",
        "      <th>score</th>\n",
        "      <th>rpkm</th>\n",
        "      <th>chromosome_id</th>\n",
        "      <th>sequence_identifier</th>\n",
        "      <th>notx_tag_count</th>\n",
        "      <th>kla_1h_tag_count</th>\n",
        "      <th>kla_6h_tag_count</th>\n",
        "      <th>kla_12h_tag_count</th>\n",
        "      <th>kla_24h_tag_count</th>\n",
        "      <th>notx_total_tags</th>\n",
        "      <th>kla_1h_total_tags</th>\n",
        "      <th>kla_6h_total_tags</th>\n",
        "      <th>kla_12h_total_tags</th>\n",
        "      <th>kla_24h_total_tags</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>  20</td>\n",
        "      <td> chr3</td>\n",
        "      <td>  7503423</td>\n",
        "      <td>  7553586</td>\n",
        "      <td> 0</td>\n",
        "      <td>  1.482268</td>\n",
        "      <td>  0.630933</td>\n",
        "      <td> 4</td>\n",
        "      <td>    NM_173181</td>\n",
        "      <td>  235</td>\n",
        "      <td> 390</td>\n",
        "      <td>  324</td>\n",
        "      <td> 290</td>\n",
        "      <td> 344</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 119</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 10012551</td>\n",
        "      <td> 10015019</td>\n",
        "      <td> 0</td>\n",
        "      <td> 71.729088</td>\n",
        "      <td> 42.748883</td>\n",
        "      <td> 4</td>\n",
        "      <td> NM_001272097</td>\n",
        "      <td> 2101</td>\n",
        "      <td> 378</td>\n",
        "      <td>  846</td>\n",
        "      <td> 982</td>\n",
        "      <td> 972</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> 150</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 14533842</td>\n",
        "      <td> 14571052</td>\n",
        "      <td> 0</td>\n",
        "      <td>  1.476441</td>\n",
        "      <td>  0.646381</td>\n",
        "      <td> 4</td>\n",
        "      <td> NM_001163580</td>\n",
        "      <td>  338</td>\n",
        "      <td> 157</td>\n",
        "      <td>  213</td>\n",
        "      <td> 242</td>\n",
        "      <td> 253</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 151</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 14578617</td>\n",
        "      <td> 14606295</td>\n",
        "      <td> 0</td>\n",
        "      <td>  1.122269</td>\n",
        "      <td>  0.505640</td>\n",
        "      <td> 4</td>\n",
        "      <td>    NM_007892</td>\n",
        "      <td>   95</td>\n",
        "      <td> 171</td>\n",
        "      <td>  214</td>\n",
        "      <td> 124</td>\n",
        "      <td>  96</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 158</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 14641702</td>\n",
        "      <td> 14663052</td>\n",
        "      <td> 0</td>\n",
        "      <td>  6.261881</td>\n",
        "      <td>  2.895446</td>\n",
        "      <td> 4</td>\n",
        "      <td>    NM_024495</td>\n",
        "      <td>  179</td>\n",
        "      <td> 591</td>\n",
        "      <td> 1166</td>\n",
        "      <td> 692</td>\n",
        "      <td> 464</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 21,
       "text": [
        "    id chr_name  transcription_start  transcription_end  strand      score  \\\n",
        "0   20     chr3              7503423            7553586       0   1.482268   \n",
        "1  119     chr3             10012551           10015019       0  71.729088   \n",
        "2  150     chr3             14533842           14571052       0   1.476441   \n",
        "3  151     chr3             14578617           14606295       0   1.122269   \n",
        "4  158     chr3             14641702           14663052       0   6.261881   \n",
        "\n",
        "        rpkm  chromosome_id sequence_identifier  notx_tag_count  \\\n",
        "0   0.630933              4           NM_173181             235   \n",
        "1  42.748883              4        NM_001272097            2101   \n",
        "2   0.646381              4        NM_001163580             338   \n",
        "3   0.505640              4           NM_007892              95   \n",
        "4   2.895446              4           NM_024495             179   \n",
        "\n",
        "   kla_1h_tag_count  kla_6h_tag_count  kla_12h_tag_count  kla_24h_tag_count  \\\n",
        "0               390               324                290                344   \n",
        "1               378               846                982                972   \n",
        "2               157               213                242                253   \n",
        "3               171               214                124                 96   \n",
        "4               591              1166                692                464   \n",
        "\n",
        "   notx_total_tags  kla_1h_total_tags  kla_6h_total_tags  kla_12h_total_tags  \\\n",
        "0         10496748            9008778           11240697            10011766   \n",
        "1         10496748            9008778           11240697            10011766   \n",
        "2         10496748            9008778           11240697            10011766   \n",
        "3         10496748            9008778           11240697            10011766   \n",
        "4         10496748            9008778           11240697            10011766   \n",
        "\n",
        "   kla_24h_total_tags  \n",
        "0             9257648  \n",
        "1             9257648  \n",
        "2             9257648  \n",
        "3             9257648  \n",
        "4             9257648  "
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "nongene_sql = '''\n",
      "SELECT\n",
      "DISTINCT ON(t.id)\n",
      "t.id,\n",
      "chr.name as chr_name,\n",
      "t.transcription_start,\n",
      "t.transcription_end,\n",
      "t.strand,\n",
      "t.score,\n",
      "t.rpkm,\n",
      "t.chromosome_id,\n",
      "notx.tag_count as notx_tag_count,\n",
      "kla_1h.tag_count as kla_1h_tag_count,\n",
      "kla_6h.tag_count as kla_6h_tag_count,\n",
      "kla_12h.tag_count as kla_12h_tag_count,\n",
      "kla_24h.tag_count as kla_24h_tag_count,\n",
      "notx_run.total_tags as notx_total_tags,\n",
      "kla_1h_run.total_tags as kla_1h_total_tags,\n",
      "kla_6h_run.total_tags as kla_6h_total_tags,\n",
      "kla_12h_run.total_tags as kla_12h_total_tags,\n",
      "kla_24h_run.total_tags as kla_24h_total_tags\n",
      "\n",
      "FROM atlas_mm10_default.atlas_transcript t\n",
      "join genome_reference_mm10.chromosome chr\n",
      "on t.chromosome_id = chr.id\n",
      "\n",
      "-- Tag counts\n",
      "join atlas_mm10_default.atlas_transcript_source notx\n",
      "on t.id = notx.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run notx_run\n",
      "on notx.sequencing_run_id = notx_run.id\n",
      "and notx_run.source_table = 'groseq\".\"tag_thiomac_gro_notx_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_1h\n",
      "on t.id = kla_1h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_1h_run\n",
      "on kla_1h.sequencing_run_id = kla_1h_run.id\n",
      "and kla_1h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_1h_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_6h\n",
      "on t.id = kla_6h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_6h_run\n",
      "on kla_6h.sequencing_run_id = kla_6h_run.id\n",
      "and kla_6h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_6h_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_12h\n",
      "on t.id = kla_12h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_12h_run\n",
      "on kla_12h.sequencing_run_id = kla_12h_run.id\n",
      "and kla_12h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_12h_1'\n",
      "\n",
      "join atlas_mm10_default.atlas_transcript_source kla_24h\n",
      "on t.id = kla_24h.atlas_transcript_id\n",
      "join genome_reference_mm10.sequencing_run kla_24h_run\n",
      "on kla_24h.sequencing_run_id = kla_24h_run.id\n",
      "and kla_24h_run.source_table = 'groseq\".\"tag_thiomac_gro_kla_24h_1'\n",
      "\n",
      "where\n",
      "t.refseq = false\n",
      "and t.score >= 1\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data = dataframe_from_query(nongene_sql, engine)\n",
      "\n",
      "data.to_csv('data/nongenes_kla.txt', sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 23
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>id</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>transcription_start</th>\n",
        "      <th>transcription_end</th>\n",
        "      <th>strand</th>\n",
        "      <th>score</th>\n",
        "      <th>rpkm</th>\n",
        "      <th>chromosome_id</th>\n",
        "      <th>notx_tag_count</th>\n",
        "      <th>kla_1h_tag_count</th>\n",
        "      <th>kla_6h_tag_count</th>\n",
        "      <th>kla_12h_tag_count</th>\n",
        "      <th>kla_24h_tag_count</th>\n",
        "      <th>notx_total_tags</th>\n",
        "      <th>kla_1h_total_tags</th>\n",
        "      <th>kla_6h_total_tags</th>\n",
        "      <th>kla_12h_total_tags</th>\n",
        "      <th>kla_24h_total_tags</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> 10</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 5576303</td>\n",
        "      <td> 5578014</td>\n",
        "      <td> 0</td>\n",
        "      <td>     5.829831</td>\n",
        "      <td>     3.667077</td>\n",
        "      <td> 4</td>\n",
        "      <td>    89</td>\n",
        "      <td>    40</td>\n",
        "      <td>    34</td>\n",
        "      <td>    62</td>\n",
        "      <td>    89</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 13</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 5860530</td>\n",
        "      <td> 5860825</td>\n",
        "      <td> 0</td>\n",
        "      <td> 17159.428223</td>\n",
        "      <td> 17312.896376</td>\n",
        "      <td> 4</td>\n",
        "      <td> 54855</td>\n",
        "      <td> 36237</td>\n",
        "      <td> 49943</td>\n",
        "      <td> 57003</td>\n",
        "      <td> 58273</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> 21</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 7553877</td>\n",
        "      <td> 7555617</td>\n",
        "      <td> 0</td>\n",
        "      <td>     1.061659</td>\n",
        "      <td>     0.666075</td>\n",
        "      <td> 4</td>\n",
        "      <td>    13</td>\n",
        "      <td>     8</td>\n",
        "      <td>    11</td>\n",
        "      <td>    12</td>\n",
        "      <td>    14</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 31</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 7612700</td>\n",
        "      <td> 7638719</td>\n",
        "      <td> 0</td>\n",
        "      <td>     9.424599</td>\n",
        "      <td>     4.272300</td>\n",
        "      <td> 4</td>\n",
        "      <td>    16</td>\n",
        "      <td>  1409</td>\n",
        "      <td>  2244</td>\n",
        "      <td>  1182</td>\n",
        "      <td>   709</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 32</td>\n",
        "      <td> chr3</td>\n",
        "      <td> 7639258</td>\n",
        "      <td> 7654062</td>\n",
        "      <td> 0</td>\n",
        "      <td>     4.516108</td>\n",
        "      <td>     2.168859</td>\n",
        "      <td> 4</td>\n",
        "      <td>     9</td>\n",
        "      <td>   337</td>\n",
        "      <td>   674</td>\n",
        "      <td>   351</td>\n",
        "      <td>   235</td>\n",
        "      <td> 10496748</td>\n",
        "      <td> 9008778</td>\n",
        "      <td> 11240697</td>\n",
        "      <td> 10011766</td>\n",
        "      <td> 9257648</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 24,
       "text": [
        "   id chr_name  transcription_start  transcription_end  strand         score  \\\n",
        "0  10     chr3              5576303            5578014       0      5.829831   \n",
        "1  13     chr3              5860530            5860825       0  17159.428223   \n",
        "2  21     chr3              7553877            7555617       0      1.061659   \n",
        "3  31     chr3              7612700            7638719       0      9.424599   \n",
        "4  32     chr3              7639258            7654062       0      4.516108   \n",
        "\n",
        "           rpkm  chromosome_id  notx_tag_count  kla_1h_tag_count  \\\n",
        "0      3.667077              4              89                40   \n",
        "1  17312.896376              4           54855             36237   \n",
        "2      0.666075              4              13                 8   \n",
        "3      4.272300              4              16              1409   \n",
        "4      2.168859              4               9               337   \n",
        "\n",
        "   kla_6h_tag_count  kla_12h_tag_count  kla_24h_tag_count  notx_total_tags  \\\n",
        "0                34                 62                 89         10496748   \n",
        "1             49943              57003              58273         10496748   \n",
        "2                11                 12                 14         10496748   \n",
        "3              2244               1182                709         10496748   \n",
        "4               674                351                235         10496748   \n",
        "\n",
        "   kla_1h_total_tags  kla_6h_total_tags  kla_12h_total_tags  \\\n",
        "0            9008778           11240697            10011766   \n",
        "1            9008778           11240697            10011766   \n",
        "2            9008778           11240697            10011766   \n",
        "3            9008778           11240697            10011766   \n",
        "4            9008778           11240697            10011766   \n",
        "\n",
        "   kla_24h_total_tags  \n",
        "0             9257648  \n",
        "1             9257648  \n",
        "2             9257648  \n",
        "3             9257648  \n",
        "4             9257648  "
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}