{
 "metadata": {
  "name": "",
  "signature": "sha256:c8c60482d8e4a37ac839f3f6c6298af57cbfe90ea37ca5e92aeaedebbe16a4be"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import OrderedDict\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas.io.sql import read_sql_query\n",
      "from sqlalchemy import create_engine\n",
      "\n",
      "\n",
      "def get_engine(uri, password, username='vespucci_user', dbname='vespucci'):\n",
      "    '''\n",
      "    dialect+driver://username:password@host:port/database\n",
      "    '''\n",
      "    return create_engine('postgresql://{}:{}@{}:5432/{}'.format(\n",
      "        username, password, uri, dbname))\n",
      "\n",
      "\n",
      "def dataframe_from_query(query, engine):\n",
      "    '''\n",
      "    Given the database connection and query, return a Pandas dataframe\n",
      "    with the requested data.\n",
      "    '''\n",
      "    return read_sql_query(query, engine)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "engine = get_engine(\n",
      "    uri='ec2-54-160-104-39.compute-1.amazonaws.com',\n",
      "    password='vespucci 2015')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ab = 'h3k4me2'\n",
      "peptides = ['no_peptide', '10um_k99a', '100um_k99a', '0_1um_pcc', '10um_pcc']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sql = '''select distinct on (p1.id) \n",
      "p1.id, chr.\"name\" as chr_name, p1.start, p1.\"end\",\n",
      "p1.chromosome_id, p1.tag_count as tag_count_{peptide},\n",
      "p2.tag_count as tag_count_{other_1},\n",
      "p2.id as id_{other_1},\n",
      "p3.tag_count as tag_count_{other_2},\n",
      "p3.id as id_{other_2},\n",
      "p4.tag_count as tag_count_{other_3},\n",
      "p4.id as id_{other_3},\n",
      "p5.tag_count as tag_count_{other_4},\n",
      "p5.id as id_{other_4}\n",
      "from chipseq.peak_cd4tcell_{ab}_{peptide}_24h_chip1 p1\n",
      "join genome_reference_mm10.chromosome chr \n",
      "on p1.chromosome_id = chr.id\n",
      "\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_1}_24h_chip1 p2\n",
      "on p1.chromosome_id = p2.chromosome_id\n",
      "and p1.start_end && p2.start_end\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_2}_24h_chip1 p3\n",
      "on p1.chromosome_id = p3.chromosome_id\n",
      "and p1.start_end && p3.start_end\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_3}_24h_chip1 p4\n",
      "on p1.chromosome_id = p4.chromosome_id\n",
      "and p1.start_end && p4.start_end\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_4}_24h_chip1 p5\n",
      "on p1.chromosome_id = p5.chromosome_id\n",
      "and p1.start_end && p5.start_end\n",
      "\n",
      "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
      "on p1.chromosome_id = reg.chromosome_id\n",
      "and p1.start_end && reg.start_site_1000\n",
      "\n",
      "where reg.id is {genic} NULL;\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 42
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "others_labels = ['other_1', 'other_2', 'other_3', 'other_4']\n",
      "for i, current in enumerate(peptides):\n",
      "    others = peptides[:i] + peptides[i+1:]\n",
      "    others_d = OrderedDict(zip(others_labels, others))\n",
      "    \n",
      "    nongene_sql = sql.format(ab=ab, peptide=current, genic='', **others_d)\n",
      "    gene_sql = sql.format(ab=ab, peptide=current, genic='NOT', **others_d)\n",
      "    \n",
      "    data = dataframe_from_query(nongene_sql, engine)\n",
      "    data.to_csv('data/me2_and/enhancers_me2_and_{}.txt'.format(current), \n",
      "                sep='\\t', header=True, index=False)\n",
      "    data = dataframe_from_query(gene_sql, engine)\n",
      "    data.to_csv('data/me2_and/promoters_me2_and_{}.txt'.format(current),\n",
      "                sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 43
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>id</th>\n",
        "      <th>chromosome_id</th>\n",
        "      <th>start</th>\n",
        "      <th>end</th>\n",
        "      <th>strand</th>\n",
        "      <th>start_end</th>\n",
        "      <th>length</th>\n",
        "      <th>summit</th>\n",
        "      <th>tag_count</th>\n",
        "      <th>raw_tag_count</th>\n",
        "      <th>...</th>\n",
        "      <th>fdr_threshold_exp</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>tag_count_no_peptide</th>\n",
        "      <th>id_no_peptide</th>\n",
        "      <th>tag_count_10um_k99a</th>\n",
        "      <th>id_10um_k99a</th>\n",
        "      <th>tag_count_100um_k99a</th>\n",
        "      <th>id_100um_k99a</th>\n",
        "      <th>tag_count_0_1um_pcc</th>\n",
        "      <th>id_0_1um_pcc</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>  5</td>\n",
        "      <td> 13</td>\n",
        "      <td>  22838517</td>\n",
        "      <td>  22859099</td>\n",
        "      <td> 0</td>\n",
        "      <td>   NumericRange(22838517, 22859099, '[)')</td>\n",
        "      <td> 20582</td>\n",
        "      <td> None</td>\n",
        "      <td> 1045.5</td>\n",
        "      <td> None</td>\n",
        "      <td>...</td>\n",
        "      <td> None</td>\n",
        "      <td> chr11</td>\n",
        "      <td>  475.6</td>\n",
        "      <td>  207</td>\n",
        "      <td> 1217.1</td>\n",
        "      <td>   6</td>\n",
        "      <td> 830.2</td>\n",
        "      <td>   8</td>\n",
        "      <td> 1213.3</td>\n",
        "      <td>  5</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td>  8</td>\n",
        "      <td> 16</td>\n",
        "      <td> 101252619</td>\n",
        "      <td> 101272613</td>\n",
        "      <td> 0</td>\n",
        "      <td> NumericRange(101252619, 101272613, '[)')</td>\n",
        "      <td> 19994</td>\n",
        "      <td> None</td>\n",
        "      <td>  887.8</td>\n",
        "      <td> None</td>\n",
        "      <td>...</td>\n",
        "      <td> None</td>\n",
        "      <td> chr15</td>\n",
        "      <td>  226.9</td>\n",
        "      <td> 1163</td>\n",
        "      <td>  824.8</td>\n",
        "      <td>  33</td>\n",
        "      <td> 342.7</td>\n",
        "      <td> 157</td>\n",
        "      <td> 1043.5</td>\n",
        "      <td> 11</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td>  9</td>\n",
        "      <td> 14</td>\n",
        "      <td>  30746035</td>\n",
        "      <td>  30764187</td>\n",
        "      <td> 0</td>\n",
        "      <td>   NumericRange(30746035, 30764187, '[)')</td>\n",
        "      <td> 18152</td>\n",
        "      <td> None</td>\n",
        "      <td>  839.9</td>\n",
        "      <td> None</td>\n",
        "      <td>...</td>\n",
        "      <td> None</td>\n",
        "      <td> chr13</td>\n",
        "      <td>  326.2</td>\n",
        "      <td>  529</td>\n",
        "      <td>  354.5</td>\n",
        "      <td> 412</td>\n",
        "      <td> 219.4</td>\n",
        "      <td> 531</td>\n",
        "      <td>  945.9</td>\n",
        "      <td> 18</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 10</td>\n",
        "      <td> 12</td>\n",
        "      <td> 109049145</td>\n",
        "      <td> 109060763</td>\n",
        "      <td> 0</td>\n",
        "      <td> NumericRange(109049145, 109060763, '[)')</td>\n",
        "      <td> 11618</td>\n",
        "      <td> None</td>\n",
        "      <td>  814.5</td>\n",
        "      <td> None</td>\n",
        "      <td>...</td>\n",
        "      <td> None</td>\n",
        "      <td>  chr9</td>\n",
        "      <td> 1081.3</td>\n",
        "      <td>   11</td>\n",
        "      <td>  985.8</td>\n",
        "      <td>  14</td>\n",
        "      <td> 704.7</td>\n",
        "      <td>  13</td>\n",
        "      <td> 1001.5</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 11</td>\n",
        "      <td>  1</td>\n",
        "      <td>  91051660</td>\n",
        "      <td>  91065868</td>\n",
        "      <td> 0</td>\n",
        "      <td>   NumericRange(91051660, 91065868, '[)')</td>\n",
        "      <td> 14208</td>\n",
        "      <td> None</td>\n",
        "      <td>  813.3</td>\n",
        "      <td> None</td>\n",
        "      <td>...</td>\n",
        "      <td> None</td>\n",
        "      <td>  chr1</td>\n",
        "      <td> 1009.5</td>\n",
        "      <td>   16</td>\n",
        "      <td>  936.1</td>\n",
        "      <td>  20</td>\n",
        "      <td> 722.8</td>\n",
        "      <td>  12</td>\n",
        "      <td> 1007.1</td>\n",
        "      <td> 14</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 26 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 32,
       "text": [
        "   id  chromosome_id      start        end  strand  \\\n",
        "0   5             13   22838517   22859099       0   \n",
        "1   8             16  101252619  101272613       0   \n",
        "2   9             14   30746035   30764187       0   \n",
        "3  10             12  109049145  109060763       0   \n",
        "4  11              1   91051660   91065868       0   \n",
        "\n",
        "                                  start_end  length summit  tag_count  \\\n",
        "0    NumericRange(22838517, 22859099, '[)')   20582   None     1045.5   \n",
        "1  NumericRange(101252619, 101272613, '[)')   19994   None      887.8   \n",
        "2    NumericRange(30746035, 30764187, '[)')   18152   None      839.9   \n",
        "3  NumericRange(109049145, 109060763, '[)')   11618   None      814.5   \n",
        "4    NumericRange(91051660, 91065868, '[)')   14208   None      813.3   \n",
        "\n",
        "  raw_tag_count      ...       fdr_threshold_exp  chr_name  \\\n",
        "0          None      ...                    None     chr11   \n",
        "1          None      ...                    None     chr15   \n",
        "2          None      ...                    None     chr13   \n",
        "3          None      ...                    None      chr9   \n",
        "4          None      ...                    None      chr1   \n",
        "\n",
        "   tag_count_no_peptide id_no_peptide  tag_count_10um_k99a id_10um_k99a  \\\n",
        "0                 475.6           207               1217.1            6   \n",
        "1                 226.9          1163                824.8           33   \n",
        "2                 326.2           529                354.5          412   \n",
        "3                1081.3            11                985.8           14   \n",
        "4                1009.5            16                936.1           20   \n",
        "\n",
        "  tag_count_100um_k99a id_100um_k99a  tag_count_0_1um_pcc  id_0_1um_pcc  \n",
        "0                830.2             8               1213.3             5  \n",
        "1                342.7           157               1043.5            11  \n",
        "2                219.4           531                945.9            18  \n",
        "3                704.7            13               1001.5            15  \n",
        "4                722.8            12               1007.1            14  \n",
        "\n",
        "[5 rows x 26 columns]"
       ]
      }
     ],
     "prompt_number": 32
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# We want one unified set of all the peaks,\n",
      "# but there is overlap between each of our sets.\n",
      "# Take the record from the set with the biggest peak\n",
      "# in each case where there are duplicates.\n",
      "# This is not perfect, especially for the cases\n",
      "# where one large peak covers several smaller,\n",
      "# but it gives us a place to start.\n",
      "\n",
      "keep_dfs = []\n",
      "tie_dfs = []\n",
      "for i, pep in enumerate(peptides):\n",
      "    data = pd.io.parsers.read_csv('data/me2_and/enhancers_me2_and_{}.txt'.format(pep), \n",
      "                                sep='\\t', header=0)\n",
      "    data = data.fillna(0)\n",
      "    print(len(data))\n",
      "    key = 'tag_count_' + pep\n",
      "    \n",
      "    peptides_before = ['tag_count_' + o for o in peptides[:i]]\n",
      "    peptides_after = ['tag_count_' + o for o in peptides[i+1:]]\n",
      "    others = peptides_before + peptides_after\n",
      "    data_mask = data[key] > data[others].max(axis=1)\n",
      "    print(sum(data_mask))\n",
      "    \n",
      "    # If there is a tie, only take this row if it is the first instance of the tie.\n",
      "    tie_mask = data[key] == data[others].max(axis=1)\n",
      "    # Is this the first instance? Fill nulls for the case of the first column.\n",
      "    first_mask = data[key] > data[peptides_before].max(axis=1).fillna(0)\n",
      "    print(sum(tie_mask & first_mask))\n",
      "    keep_dfs.append(data[data_mask | (tie_mask & first_mask)])\n",
      "    \n",
      "keep = pd.concat(keep_dfs, join='outer', axis=0, ignore_index=True)\n",
      "print(len(keep))\n",
      "keep.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "19472\n",
        "8924\n",
        "54\n",
        "20586"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "6701\n",
        "30\n",
        "19871"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "4488\n",
        "14\n",
        "21207"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "8794\n",
        "5\n",
        "21078"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "6546\n",
        "0\n",
        "35556\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chr_name</th>\n",
        "      <th>chromosome_id</th>\n",
        "      <th>end</th>\n",
        "      <th>id</th>\n",
        "      <th>id_0_1um_pcc</th>\n",
        "      <th>id_100um_k99a</th>\n",
        "      <th>id_10um_k99a</th>\n",
        "      <th>id_10um_pcc</th>\n",
        "      <th>id_no_peptide</th>\n",
        "      <th>start</th>\n",
        "      <th>tag_count_0_1um_pcc</th>\n",
        "      <th>tag_count_100um_k99a</th>\n",
        "      <th>tag_count_10um_k99a</th>\n",
        "      <th>tag_count_10um_pcc</th>\n",
        "      <th>tag_count_no_peptide</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>  chr6</td>\n",
        "      <td>  7</td>\n",
        "      <td>  41548036</td>\n",
        "      <td>  5</td>\n",
        "      <td>   8</td>\n",
        "      <td>    41</td>\n",
        "      <td>    8</td>\n",
        "      <td>   53</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  41532088</td>\n",
        "      <td> 1105.9</td>\n",
        "      <td> 531.1</td>\n",
        "      <td> 1133.6</td>\n",
        "      <td> 561.5</td>\n",
        "      <td> 1248.2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td>  chr7</td>\n",
        "      <td>  8</td>\n",
        "      <td>  83656974</td>\n",
        "      <td> 12</td>\n",
        "      <td>  37</td>\n",
        "      <td>    22</td>\n",
        "      <td>   26</td>\n",
        "      <td>   34</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  83645498</td>\n",
        "      <td>  823.1</td>\n",
        "      <td> 610.8</td>\n",
        "      <td>  880.5</td>\n",
        "      <td> 594.4</td>\n",
        "      <td> 1046.2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td>  chr2</td>\n",
        "      <td>  2</td>\n",
        "      <td>  32546809</td>\n",
        "      <td> 26</td>\n",
        "      <td>  42</td>\n",
        "      <td> 17822</td>\n",
        "      <td>   49</td>\n",
        "      <td>   51</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  32536709</td>\n",
        "      <td>  807.0</td>\n",
        "      <td>  25.4</td>\n",
        "      <td>  751.8</td>\n",
        "      <td> 566.1</td>\n",
        "      <td>  923.6</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td>  chr8</td>\n",
        "      <td> 10</td>\n",
        "      <td>  72367600</td>\n",
        "      <td> 30</td>\n",
        "      <td> 344</td>\n",
        "      <td> 20180</td>\n",
        "      <td> 3631</td>\n",
        "      <td> 6527</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  72348242</td>\n",
        "      <td>  385.9</td>\n",
        "      <td>  20.9</td>\n",
        "      <td>  138.2</td>\n",
        "      <td>  78.0</td>\n",
        "      <td>  910.3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> chr12</td>\n",
        "      <td> 15</td>\n",
        "      <td> 108001601</td>\n",
        "      <td> 32</td>\n",
        "      <td>  33</td>\n",
        "      <td>    19</td>\n",
        "      <td>   30</td>\n",
        "      <td>   28</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 107989079</td>\n",
        "      <td>  854.6</td>\n",
        "      <td> 654.3</td>\n",
        "      <td>  840.3</td>\n",
        "      <td> 626.7</td>\n",
        "      <td>  891.1</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 50,
       "text": [
        "  chr_name  chromosome_id        end  id  id_0_1um_pcc  id_100um_k99a  \\\n",
        "0     chr6              7   41548036   5             8             41   \n",
        "1     chr7              8   83656974  12            37             22   \n",
        "2     chr2              2   32546809  26            42          17822   \n",
        "3     chr8             10   72367600  30           344          20180   \n",
        "4    chr12             15  108001601  32            33             19   \n",
        "\n",
        "   id_10um_k99a  id_10um_pcc  id_no_peptide      start  tag_count_0_1um_pcc  \\\n",
        "0             8           53            NaN   41532088               1105.9   \n",
        "1            26           34            NaN   83645498                823.1   \n",
        "2            49           51            NaN   32536709                807.0   \n",
        "3          3631         6527            NaN   72348242                385.9   \n",
        "4            30           28            NaN  107989079                854.6   \n",
        "\n",
        "   tag_count_100um_k99a  tag_count_10um_k99a  tag_count_10um_pcc  \\\n",
        "0                 531.1               1133.6               561.5   \n",
        "1                 610.8                880.5               594.4   \n",
        "2                  25.4                751.8               566.1   \n",
        "3                  20.9                138.2                78.0   \n",
        "4                 654.3                840.3               626.7   \n",
        "\n",
        "   tag_count_no_peptide  \n",
        "0                1248.2  \n",
        "1                1046.2  \n",
        "2                 923.6  \n",
        "3                 910.3  \n",
        "4                 891.1  "
       ]
      }
     ],
     "prompt_number": 50
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}