{
 "metadata": {
  "name": "",
  "signature": "sha256:942c251458fb321ea864601e47ca88701d40feb0dc30c0630030a2a7a046e5d3"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import OrderedDict\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas.io.sql import read_sql_query\n",
      "from sqlalchemy import create_engine\n",
      "\n",
      "\n",
      "def get_engine(uri, password, username='vespucci_user', dbname='vespucci'):\n",
      "    '''\n",
      "    dialect+driver://username:password@host:port/database\n",
      "    '''\n",
      "    return create_engine('postgresql://{}:{}@{}:5432/{}'.format(\n",
      "        username, password, uri, dbname))\n",
      "\n",
      "\n",
      "def dataframe_from_query(query, engine):\n",
      "    '''\n",
      "    Given the database connection and query, return a Pandas dataframe\n",
      "    with the requested data.\n",
      "    '''\n",
      "    return read_sql_query(query, engine)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "engine = get_engine(\n",
      "    uri='ec2-54-160-104-39.compute-1.amazonaws.com',\n",
      "    password='vespucci 2015')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ab = 'h3k27ac'\n",
      "peptides = ['no_peptide', '1um_k99a', '1um_pcc']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sql = '''select distinct on (p1.id) \n",
      "p1.id, chr.\"name\" as chr_name, p1.start, p1.\"end\",\n",
      "p1.chromosome_id, \n",
      "p1.tag_count as tag_count_{peptide},\n",
      "p1.id as id_{peptide},\n",
      "p2.tag_count as tag_count_{other_1},\n",
      "p2.id as id_{other_1},\n",
      "p3.tag_count as tag_count_{other_2},\n",
      "p3.id as id_{other_2}\n",
      "\n",
      "from chipseq.peak_cd4tcell_{ab}_{peptide}_24h p1\n",
      "join genome_reference_mm10.chromosome chr \n",
      "on p1.chromosome_id = chr.id\n",
      "\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_1}_24h p2\n",
      "on p1.chromosome_id = p2.chromosome_id\n",
      "and p1.start_end && p2.start_end\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_{other_2}_24h p3\n",
      "on p1.chromosome_id = p3.chromosome_id\n",
      "and p1.start_end && p3.start_end\n",
      "\n",
      "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
      "on p1.chromosome_id = reg.chromosome_id\n",
      "and p1.start_end && reg.start_site_1000\n",
      "\n",
      "where reg.id is {genic} NULL;\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "others_labels = ['other_1', 'other_2']\n",
      "for i, current in enumerate(peptides):\n",
      "    others = peptides[:i] + peptides[i+1:]\n",
      "    others_d = OrderedDict(zip(others_labels, others))\n",
      "    \n",
      "    nongene_sql = sql.format(ab=ab, peptide=current, genic='', **others_d)\n",
      "    gene_sql = sql.format(ab=ab, peptide=current, genic='NOT', **others_d)\n",
      "    \n",
      "    data = dataframe_from_query(nongene_sql, engine)\n",
      "    data.to_csv('data/enhancers_{}_and_{}.txt'.format(ab, current), \n",
      "                sep='\\t', header=True, index=False)\n",
      "    data = dataframe_from_query(gene_sql, engine)\n",
      "    data.to_csv('data/promoters_{}_and_{}.txt'.format(ab, current),\n",
      "                sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>id</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>start</th>\n",
        "      <th>end</th>\n",
        "      <th>chromosome_id</th>\n",
        "      <th>tag_count_1um_pcc</th>\n",
        "      <th>id_1um_pcc</th>\n",
        "      <th>tag_count_no_peptide</th>\n",
        "      <th>id_no_peptide</th>\n",
        "      <th>tag_count_1um_k99a</th>\n",
        "      <th>id_1um_k99a</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td>  2</td>\n",
        "      <td>  chr9</td>\n",
        "      <td>  32691795</td>\n",
        "      <td>  32753483</td>\n",
        "      <td> 12</td>\n",
        "      <td> 3197.6</td>\n",
        "      <td>  2</td>\n",
        "      <td> 1370.1</td>\n",
        "      <td>   3</td>\n",
        "      <td>  675.9</td>\n",
        "      <td> 119</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td>  5</td>\n",
        "      <td> chr13</td>\n",
        "      <td>  24590249</td>\n",
        "      <td>  24649011</td>\n",
        "      <td> 14</td>\n",
        "      <td> 2149.5</td>\n",
        "      <td>  5</td>\n",
        "      <td> 1334.9</td>\n",
        "      <td>   4</td>\n",
        "      <td> 1753.5</td>\n",
        "      <td>   9</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td>  7</td>\n",
        "      <td>  chr2</td>\n",
        "      <td>  29618154</td>\n",
        "      <td>  29673520</td>\n",
        "      <td>  2</td>\n",
        "      <td> 2062.1</td>\n",
        "      <td>  7</td>\n",
        "      <td>  656.3</td>\n",
        "      <td>  51</td>\n",
        "      <td> 1762.9</td>\n",
        "      <td>   8</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td>  9</td>\n",
        "      <td>  chr2</td>\n",
        "      <td>  11164377</td>\n",
        "      <td>  11212807</td>\n",
        "      <td>  2</td>\n",
        "      <td> 1795.6</td>\n",
        "      <td>  9</td>\n",
        "      <td>  437.1</td>\n",
        "      <td> 164</td>\n",
        "      <td> 2252.0</td>\n",
        "      <td>   2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 10</td>\n",
        "      <td>  chr6</td>\n",
        "      <td> 129207907</td>\n",
        "      <td> 129243211</td>\n",
        "      <td>  7</td>\n",
        "      <td> 1794.3</td>\n",
        "      <td> 10</td>\n",
        "      <td> 1119.0</td>\n",
        "      <td>   5</td>\n",
        "      <td> 1586.5</td>\n",
        "      <td>  13</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 13,
       "text": [
        "   id chr_name      start        end  chromosome_id  tag_count_1um_pcc  \\\n",
        "0   2     chr9   32691795   32753483             12             3197.6   \n",
        "1   5    chr13   24590249   24649011             14             2149.5   \n",
        "2   7     chr2   29618154   29673520              2             2062.1   \n",
        "3   9     chr2   11164377   11212807              2             1795.6   \n",
        "4  10     chr6  129207907  129243211              7             1794.3   \n",
        "\n",
        "   id_1um_pcc  tag_count_no_peptide  id_no_peptide  tag_count_1um_k99a  \\\n",
        "0           2                1370.1              3               675.9   \n",
        "1           5                1334.9              4              1753.5   \n",
        "2           7                 656.3             51              1762.9   \n",
        "3           9                 437.1            164              2252.0   \n",
        "4          10                1119.0              5              1586.5   \n",
        "\n",
        "   id_1um_k99a  \n",
        "0          119  \n",
        "1            9  \n",
        "2            8  \n",
        "3            2  \n",
        "4           13  "
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# We want one unified set of all the peaks,\n",
      "# but there is overlap between each of our sets.\n",
      "# Take the record from the set with the biggest peak\n",
      "# in each case where there are duplicates.\n",
      "# This is not perfect, especially for the cases\n",
      "# where one large peak covers several smaller,\n",
      "# but it gives us a place to start.\n",
      "\n",
      "for peak_type in ('enhancers','promoters'):\n",
      "    keep_dfs = []\n",
      "    for i, pep in enumerate(peptides):\n",
      "        data = pd.io.parsers.read_csv('data/{}_{}_and_{}.txt'.format(peak_type, ab, pep), \n",
      "                                    sep='\\t', header=0)\n",
      "        data = data.fillna(0)\n",
      "        print(len(data))\n",
      "        key = 'tag_count_' + pep\n",
      "\n",
      "        peptides_before = ['tag_count_' + o for o in peptides[:i]]\n",
      "        peptides_after = ['tag_count_' + o for o in peptides[i+1:]]\n",
      "        others = peptides_before + peptides_after\n",
      "        data_mask = data[key] > data[others].max(axis=1)\n",
      "        print(sum(data_mask))\n",
      "\n",
      "        # If there is a tie, only take this row if it is the first instance of the tie.\n",
      "        tie_mask = data[key] == data[others].max(axis=1)\n",
      "        # Is this the first instance? Fill nulls for the case of the first column.\n",
      "        first_mask = data[key] > data[peptides_before].max(axis=1).fillna(0)\n",
      "        print(sum(tie_mask & first_mask))\n",
      "        keep_dfs.append(data[data_mask | (tie_mask & first_mask)])\n",
      "\n",
      "    keep = pd.concat(keep_dfs, join='outer', axis=0, ignore_index=True)\n",
      "    print(len(keep))\n",
      "    keep.head()\n",
      "    keep.to_csv('data/{}_{}_and_union.txt'.format(peak_type, ab), \n",
      "                sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "8309\n",
        "1319\n",
        "11\n",
        "13737\n",
        "8032\n",
        "8\n",
        "11450"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "5826\n",
        "0\n",
        "15196\n",
        "7124"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "537\n",
        "3\n",
        "8752\n",
        "2637\n",
        "13\n",
        "8851\n",
        "6059\n",
        "0\n",
        "9249"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}