{
 "metadata": {
  "name": "",
  "signature": "sha256:3dec3de1ef6da12f36dd5179c83f0920b8a18ec145c7046729b737cf9676c84b"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "Get data for No-Peptide replicates for scatterplot.\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1,
       "text": [
        "'\\nGet data for No-Peptide replicates for scatterplot.\\n'"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import OrderedDict\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas.io.sql import read_sql_query\n",
      "from sqlalchemy import create_engine\n",
      "\n",
      "\n",
      "def get_engine(uri, password, username='vespucci_user', dbname='vespucci'):\n",
      "    '''\n",
      "    dialect+driver://username:password@host:port/database\n",
      "    '''\n",
      "    return create_engine('postgresql://{}:{}@{}:5432/{}'.format(\n",
      "        username, password, uri, dbname))\n",
      "\n",
      "\n",
      "def dataframe_from_query(query, engine):\n",
      "    '''\n",
      "    Given the database connection and query, return a Pandas dataframe\n",
      "    with the requested data.\n",
      "    '''\n",
      "    return read_sql_query(query, engine)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "engine = get_engine(\n",
      "    uri='ec2-54-160-104-39.compute-1.amazonaws.com',\n",
      "    password='vespucci 2015')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ab = 'h3k4me2'\n",
      "dates = ['13_11_20', '13_11_25', '14_06_11']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sql = '''select distinct on (p1.id) \n",
      "p1.id, chr.\"name\" as chr_name, p1.start, p1.\"end\",\n",
      "p1.chromosome_id, \n",
      "p1.tag_count as tag_count_{date},\n",
      "p1.id as id_{date},\n",
      "p2.tag_count as tag_count_{other_1},\n",
      "p2.id as id_{other_1},\n",
      "p3.tag_count as tag_count_{other_2},\n",
      "p3.id as id_{other_2}\n",
      "\n",
      "from chipseq.peak_cd4tcell_{ab}_no_peptide_24h_{date} p1\n",
      "join genome_reference_mm10.chromosome chr \n",
      "on p1.chromosome_id = chr.id\n",
      "\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_no_peptide_24h_{other_1} p2\n",
      "on p1.chromosome_id = p2.chromosome_id\n",
      "and p1.start_end && p2.start_end\n",
      "left outer join chipseq.peak_cd4tcell_{ab}_no_peptide_24h_{other_2} p3\n",
      "on p1.chromosome_id = p3.chromosome_id\n",
      "and p1.start_end && p3.start_end\n",
      "\n",
      "left outer join genome_reference_mm10.sequence_transcription_region reg\n",
      "on p1.chromosome_id = reg.chromosome_id\n",
      "and p1.start_end && reg.start_site_1000\n",
      "\n",
      "where reg.id is {genic} NULL;\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "others_labels = ['other_1', 'other_2']\n",
      "for i, current in enumerate(dates):\n",
      "    others = dates[:i] + dates[i+1:]\n",
      "    others_d = OrderedDict(zip(others_labels, others))\n",
      "    \n",
      "    nongene_sql = sql.format(ab=ab, date=current, genic='', **others_d)\n",
      "    gene_sql = sql.format(ab=ab, date=current, genic='NOT', **others_d)\n",
      "    \n",
      "    data = dataframe_from_query(nongene_sql, engine)\n",
      "    data.to_csv('data/enhancers_me2_and_no_peptide_{}.txt'.format(current), \n",
      "                sep='\\t', header=True, index=False)\n",
      "    data = dataframe_from_query(gene_sql, engine)\n",
      "    data.to_csv('data/promoters_me2_and_no_peptide_{}.txt'.format(current),\n",
      "                sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>id</th>\n",
        "      <th>chr_name</th>\n",
        "      <th>start</th>\n",
        "      <th>end</th>\n",
        "      <th>chromosome_id</th>\n",
        "      <th>tag_count_14_06_11</th>\n",
        "      <th>id_14_06_11</th>\n",
        "      <th>tag_count_13_11_20</th>\n",
        "      <th>id_13_11_20</th>\n",
        "      <th>tag_count_13_11_25</th>\n",
        "      <th>id_13_11_25</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> 2</td>\n",
        "      <td>  chr9</td>\n",
        "      <td>  32692621</td>\n",
        "      <td>  32734051</td>\n",
        "      <td> 12</td>\n",
        "      <td> 3598.7</td>\n",
        "      <td> 2</td>\n",
        "      <td> 1552.6</td>\n",
        "      <td>   2</td>\n",
        "      <td> 913.0</td>\n",
        "      <td>    2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 3</td>\n",
        "      <td> chr11</td>\n",
        "      <td>  11685952</td>\n",
        "      <td>  11742900</td>\n",
        "      <td> 13</td>\n",
        "      <td> 3593.9</td>\n",
        "      <td> 3</td>\n",
        "      <td>  629.1</td>\n",
        "      <td>  59</td>\n",
        "      <td> 584.8</td>\n",
        "      <td>   17</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> 4</td>\n",
        "      <td>  chr1</td>\n",
        "      <td> 138129824</td>\n",
        "      <td> 138176314</td>\n",
        "      <td>  1</td>\n",
        "      <td> 3114.9</td>\n",
        "      <td> 4</td>\n",
        "      <td>  476.4</td>\n",
        "      <td> 146</td>\n",
        "      <td> 273.1</td>\n",
        "      <td>  324</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 5</td>\n",
        "      <td>  chr3</td>\n",
        "      <td>  89378218</td>\n",
        "      <td>  89397222</td>\n",
        "      <td>  4</td>\n",
        "      <td> 2939.3</td>\n",
        "      <td> 5</td>\n",
        "      <td>  863.9</td>\n",
        "      <td>  15</td>\n",
        "      <td> 647.0</td>\n",
        "      <td>   12</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 6</td>\n",
        "      <td>  chr5</td>\n",
        "      <td>  65855826</td>\n",
        "      <td>  65890258</td>\n",
        "      <td>  6</td>\n",
        "      <td> 2925.9</td>\n",
        "      <td> 6</td>\n",
        "      <td>  875.4</td>\n",
        "      <td>  14</td>\n",
        "      <td>  78.8</td>\n",
        "      <td> 8065</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 18,
       "text": [
        "   id chr_name      start        end  chromosome_id  tag_count_14_06_11  \\\n",
        "0   2     chr9   32692621   32734051             12              3598.7   \n",
        "1   3    chr11   11685952   11742900             13              3593.9   \n",
        "2   4     chr1  138129824  138176314              1              3114.9   \n",
        "3   5     chr3   89378218   89397222              4              2939.3   \n",
        "4   6     chr5   65855826   65890258              6              2925.9   \n",
        "\n",
        "   id_14_06_11  tag_count_13_11_20  id_13_11_20  tag_count_13_11_25  \\\n",
        "0            2              1552.6            2               913.0   \n",
        "1            3               629.1           59               584.8   \n",
        "2            4               476.4          146               273.1   \n",
        "3            5               863.9           15               647.0   \n",
        "4            6               875.4           14                78.8   \n",
        "\n",
        "   id_13_11_25  \n",
        "0            2  \n",
        "1           17  \n",
        "2          324  \n",
        "3           12  \n",
        "4         8065  "
       ]
      }
     ],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# We want one unified set of all the peaks,\n",
      "# but there is overlap between each of our sets.\n",
      "# Take the record from the set with the biggest peak\n",
      "# in each case where there are duplicates.\n",
      "# This is not perfect, especially for the cases\n",
      "# where one large peak covers several smaller,\n",
      "# but it gives us a place to start.\n",
      "\n",
      "for peak_type in ('enhancers','promoters'):\n",
      "    keep_dfs = []\n",
      "    for i, date in enumerate(dates):\n",
      "        data = pd.io.parsers.read_csv('data/{}_me2_and_no_peptide_{}.txt'.format(peak_type, date), \n",
      "                                    sep='\\t', header=0)\n",
      "        data = data.fillna(0)\n",
      "        print(len(data))\n",
      "        key = 'tag_count_' + date\n",
      "\n",
      "        before = ['tag_count_' + o for o in dates[:i]]\n",
      "        after = ['tag_count_' + o for o in dates[i+1:]]\n",
      "        others = before + after\n",
      "        data_mask = data[key] > data[others].max(axis=1)\n",
      "        print(sum(data_mask))\n",
      "\n",
      "        # If there is a tie, only take this row if it is the first instance of the tie.\n",
      "        tie_mask = data[key] == data[others].max(axis=1)\n",
      "        # Is this the first instance? Fill nulls for the case of the first column.\n",
      "        first_mask = data[key] > data[before].max(axis=1).fillna(0)\n",
      "        print(sum(tie_mask & first_mask))\n",
      "        keep_dfs.append(data[data_mask | (tie_mask & first_mask)])\n",
      "\n",
      "    keep = pd.concat(keep_dfs, join='outer', axis=0, ignore_index=True)\n",
      "    print(len(keep))\n",
      "    keep.head()\n",
      "    keep.to_csv('data/{}_me2_and_union_no_peptide.txt'.format(peak_type), \n",
      "                sep='\\t', header=True, index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "39147\n",
        "15705\n",
        "49\n",
        "31190\n",
        "8070\n",
        "15"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "41974"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "34564\n",
        "0\n",
        "58403\n",
        "12977"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "751\n",
        "1\n",
        "12224\n",
        "195\n",
        "0\n",
        "13875"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "13044\n",
        "0\n",
        "13991\n"
       ]
      }
     ],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}